File: benchmarks/tsb/bench_to_from_dict.ts
/**
 * Benchmark: toDictOriented / fromDictOriented — DataFrame ↔ dict conversions.
 * Tests all orient variants: "list", "records", "split", "index", "tight".
 * Outputs JSON: {"function": "to_from_dict", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, toDictOriented, fromDictOriented } from "../../src/index.js";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 50;

const df = new DataFrame({
  a: Array.from({ length: SIZE }, (_, i) => i),
  b: Array.from({ length: SIZE }, (_, i) => i * 1.5),
  c: Array.from({ length: SIZE }, (_, i) => `str_${i % 100}`),
});

for (let i = 0; i < WARMUP; i++) {
  toDictOriented(df, "list");
  toDictOriented(df, "records");
  toDictOriented(df, "split");
  toDictOriented(df, "index");
  toDictOriented(df, "tight");
  fromDictOriented({ a: [1, 2, 3], b: [4, 5, 6] });
  fromDictOriented({ 0: { a: 1, b: 4 }, 1: { a: 2, b: 5 } }, "index");
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  toDictOriented(df, "list");
  toDictOriented(df, "records");
  toDictOriented(df, "split");
  toDictOriented(df, "index");
  toDictOriented(df, "tight");
  fromDictOriented({ a: [1, 2, 3], b: [4, 5, 6] });
  fromDictOriented({ 0: { a: 1, b: 4 }, 1: { a: 2, b: 5 } }, "index");
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "to_from_dict",
    mean_ms: Math.round((total / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_to_from_dict.py
"""
Benchmark: DataFrame.to_dict / DataFrame.from_dict — dict orient conversions.
Tests: list, records, split, index orient round-trips on a 10k-row DataFrame.
Outputs JSON: {"function": "to_from_dict", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import time
import json
import pandas as pd

SIZE = 10_000
WARMUP = 5
ITERATIONS = 50

df = pd.DataFrame({
    "a": list(range(SIZE)),
    "b": [i * 1.5 for i in range(SIZE)],
    "c": [f"str_{i % 100}" for i in range(SIZE)],
})

small_list = {"a": [1, 2, 3], "b": [4, 5, 6]}
small_df = pd.DataFrame(small_list)
small_index = {0: {"a": 1, "b": 4}, 1: {"a": 2, "b": 5}}

for _ in range(WARMUP):
    df.to_dict(orient="list")
    df.to_dict(orient="records")
    df.to_dict(orient="split")
    df.to_dict(orient="index")
    pd.DataFrame.from_dict(small_list)
    pd.DataFrame.from_dict(small_index, orient="index")

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.to_dict(orient="list")
    df.to_dict(orient="records")
    df.to_dict(orient="split")
    df.to_dict(orient="index")
    pd.DataFrame.from_dict(small_list)
    pd.DataFrame.from_dict(small_index, orient="index")
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "to_from_dict",
    "mean_ms": round(total / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))
